### ISSUE TITLE
Understanding Reports

### ISSUE THREAD

**Issue by Codeusa** [nefarius/ViGEm#45]
I'm trying to grasp how sending reports to the controller is suppose to be done.

After "plugging in" the Xbox 360 controller, I generate a report like so.

```
 var report = new Xbox360Report();
                    report.SetButtons(Xbox360Buttons.A, Xbox360Buttons.B);
                    xboxController.SendReport(report);
```

However it almost always fails to register, if I place it in a loop it usually successfully registers the second time around. 

Then there is the problem with button states, i assume removing a button hold should be as simple as removing the button from the report. Which does work when I do it, however when trying to press that button in a new report again, nothing happens. 

![](https://i.imgur.com/jHHKgTA.png)

As you can see, nothing happens when a pressed report is sent. 

tl;dr

- What is the correct way to handle multiple button presses in quick succession? 
- Whats up with the hit or miss button pressing? 

C# Wrapper
Windows 10 Latest SDK

**Comment by Codeusa** [comment]
@ToadKing  I figured I'd tag you since you seem to be digging into this project.

**Comment by ToadKing** [comment]
It sounds like you might be running into this bug: https://github.com/nefarius/ViGEm/issues/4

I work around it by sending each report three times. The driver seems to handle it well enough and it's not really expensive to do it.

The ViGEmUM library has completely control over all buttons/triggers/sticks so it might just be the C# wrapper that doesn't have trigger support yet.

**Comment by Codeusa** [comment]
By sending 3 times do you not risk duplicate inputs?

**Comment by ToadKing** [comment]
XInput is poll-based so the duplicate reports never get to the program.

I'm not convinced this is a bug with xusb22.sys myself. My thought is ViGEm is not following something exactly from the XUSB spec, which unfortunately is not public so it'll be difficult to find out exactly what it's doing wrong.

**Comment by Codeusa** [comment]
Thanks for the assistance, it does seem sending 3 times fixes the issue.

**Comment by Alia5** [comment]
I dunno if I ran into the same issue before, however, instead of sending each report 3 times, I've turned the process priority to "high". Now every action submits reliably.

**Comment by JoshWobbles** [comment]
Try slowing the report speed down. Default USB pooling is only 125hz so you could likely slow it down to that without seeing any difference.

**Comment by ToadKing** [comment]
@Alia5 Does not fix the issue for me.

@JoshWobbles This doesn't fix it either. I changed it so it only sends a report once every second and the behavior in #4 still happens.

**Comment by nefarius** [comment]
I can replicate this. Sure could be it's my drivers fault but since I'm queuing and answering every request I get I assumed it's some `XUSB.SYS` internal "issue" since it also happens when you poll for vibration values: even if you bombard the bus with rumble requests only every second increase of the vibration values will be passed to the user-mode functions. I put this on low-priority for myself since usually you call the feeding functions in a loop to constantly provide changes.

**Comment by ToadKing** [comment]
That's true. I only noticed the bug when I added in a check to only send new reports when the controller state changed.

**Comment by nefarius** [comment]
Do bear in mind that XUSB expects - like the name suggests - communication with an actual USB device. Maybe the hardware-pad always sends multiple interrupts per state change, I can't recall for sure from my memory when Wiresharking them 😉

**Comment by nefarius** [comment]
@Alia5 what the...? I smell a race condition...

**Comment by Alia5** [comment]
In my case, Steam and it's hooks are always a factor too, and that's prob. the case for me then...

**Comment by catid** [comment]
Noticed some things recently:

(Thing 1) [https://github.com/nefarius/ViGEm/blob/master/ViGEmBus/busenum.c#L428](https://github.com/nefarius/ViGEm/blob/master/ViGEmBus/busenum.c#L428)
It looks like this is comparing the wrong offset since the report is at +2, which means changed is probably always true.  I'm not sure you want to fix this though because issue #4 may mean that input starts getting dropped entirely.

(Thing 2) [https://github.com/nefarius/ViGEm/blob/master/ViGEmBus/busenum.c#L462](https://github.com/nefarius/ViGEm/blob/master/ViGEmBus/busenum.c#L462)
If we do not have a request ready to go here then the input gets lost.  It seems like there should be a way to store off the latest submitted data behind a lock, and then change the code here to return data immediately:
[https://github.com/nefarius/ViGEm/blob/master/ViGEmBus/usbpdo.c#L508](https://github.com/nefarius/ViGEm/blob/master/ViGEmBus/usbpdo.c#L508).
Perhaps this is what's happening with the 3x request issue?

(Thing 3) It would be interesting to see how deep the request queue is and how it relates to input latency.  Perhaps the queue depth is 3 and the proper behavior is to retire the input sooner?

**Comment by nefarius** [comment]
@catid thanks for your input:

Ad 1) What the sh... I thought I fixed that. Must have reverted it by accident in the past. To be honest, I just migrated this check over from `ScpVBus`, it shouldn't be necessary anyways.

Ad 2) Yes, this is possible yet very unlikely. I am VERY afraid of introducing locking there because, you know, deadlock in kernel land is sudden death. But it should be refined, yes.

Ad 3) The request queue can always hold either zero or one request because the upper driver won't send any more interrupt requests before the previous one got completed.
